### DISCUSSION TITLE
Do I need to install windows-build-tools? If so, should I install it for every node version?

### DISCUSSION THREAD

**Discussion by leafac** [coreybutler/nvm-windows#627] [Support] [answered]
I’m talking about this package: https://npm.im/windows-build-tools

It installs a C/C++ compiler and Python so that you can install packages with native modules (for example, [sharp](https://npm.im/sharp)).

My questions are:

1. Does nvm subsume windows-build-tools?
2. If not, then should I treat windows-build-tools like any other global utility [and reinstall it for every version of Node](https://github.com/coreybutler/nvm-windows#reinstall-any-global-utilities), or does it suffice to install it once? (After all, windows-build-tools is **not** a command-line tool.)

I’m asking the same question over at the windows-build-tools repository: https://github.com/felixrieseberg/windows-build-tools/issues/239

**Comment by leafac** [comment, marked as answer]
The question was answered by @jacobq at https://github.com/felixrieseberg/windows-build-tools/issues/239

In summary:

**nvm does **not** subsume windows-build-tools.** You should still install windows-build-tools to work with native dependencies.

**You may install windows-build-tools only once.** Unlike other global utilities, this isn’t a command-line tool that should be available when you switch to different Node version. Once the C/C++ compiler and Python are installed, they’ll continue to work even if you switch to different Node versions.

**Comment by hhorneber** [comment]
This worked for me:

* Install node with native build tools using the installer provided by nodejs.org
* Uninstall node (this will leave the build tools in place)
* Install nvm

**Reply by leafac** [reply]
Good point.

Since writing my messages above I found that if you’re using Chocolatey it may be easier to just install the Node.js build tools through it, instead of relying on `windows-build-tools` or the Node.js installer:

```console
> choco install python visualstudio2022-workload-vctools
```

**Reply by hhorneber** [reply]
That actually looks like the best solution to me.

**Reply by Taschenbuch** [reply]
Or: Do not install node from nodejs.org. Just install nvm. use nvm install to install a nodejs version.
1. Enter this in windows file explorer path: %AppData%\nvm
2. open the folder for the version you just installed with nvm
3. double click on install_tools.bat which will install python and the vsbuild tools in the correct versions.  it will cause some UAC prompts

**Reply by angelicochris** [reply]
> Or: Do not install node from nodejs.org. Just install nvm. use nvm install to install a nodejs version.
> 
>     1. Enter this in windows file explorer path: %AppData%\nvm
> 
>     2. open the folder for the version you just installed with nvm
> 
>     3. double click on install_tools.bat which will install python and the vsbuild tools in the correct versions.  it will cause some UAC prompts

This doesn't seem to be the case, only time I was given install_tools.bat is when it was downloaded from nodejs.org.
Why doesn't nvm-windows install build tools if they are missing when its first installed?

**Reply by Taschenbuch** [reply]
may depend on your node version. For 16 and 18 there is a install_tools.bat inside the version folder. Tested on multiple systems in different networks. But i agree. I think nvm-windows should at least ask whether it should install the builds tools too like the official windows nodejs installer asks as well.
